src/box/upgrade.c

```diff
@@ -0,0 +1,419 @@
+/*
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "box.h"
+#include "func.h"
+#include "iproto_constants.h"
+#include "space.h"
+#include "tuple.h"
+#include "txn.h"
+#include "port.h"
+#include "schema.h"
+#include "upgrade.h"
+#include "memtx_space.h"
+#include "xrow.h"
+
+const char *upgrade_status_strs[] = {
+	"inprogress",
+	"error",
+	"test"
+};
+
+enum {
+	/**
+	 * Batch tuples to be upgraded in order to reduce number of yields.
+	 */
+#if defined(NDEBUG)
+	UPGRADE_TX_BATCH_SIZE = 1024,
+#else
+	UPGRADE_TX_BATCH_SIZE = 10
+#endif
+};
+
+void
+space_upgrade_delete(struct space_upgrade *upgrade)
+{
+	if (upgrade->format != NULL) {
+		assert(upgrade->format->refs == 1);
+		tuple_format_unref(upgrade->format);
+	}
+	free(upgrade);
+}
+
+/**
+ * Function is lightweight variation of box_process_rw():
+ * it fills-in ephemeral REPLACE request (it is required to process data
+ * to WAL), begins/commits statement, creates new tuple from @a new_tuple
+ * msgpack and replaces it to the all indexes.
+ */
+static int
+tuple_upgrade(struct space *space, struct tuple *old_tuple,
+	      const char *new_tuple, const char *new_tuple_end)
+{
+	assert(space_is_memtx(space));
```

Comment:
Firstly, space_execute_dml() involves sequences and triggers which in turn shouldn't be affected.
Secondly, the initial design document  (by product team) declared support only for memtx engine. I believe it's
not really a problem to run upgrade on vinyl spaces but let's delay this feature until there's
real demand for it.

Reply:
Let's at least try to implement the multi-engine test and check Vinyl. There's nothing engine-dependent in this patch. It should work out of the box for any engine. If it doesn't because of a bug in Vinyl, you can file a bug and temporarily disable the test for Vinyl.